libs/gl-client/src/pairing/attestation_device.rs

```diff
@@ -0,0 +1,265 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use super::{into_approve_pairing_error, into_verify_pairing_data_error, Error};
+use crate::{
+    credentials::{RuneProvider, TlsConfigProvider},
+    pb::scheduler::{
+        pairing_client::PairingClient, ApprovePairingRequest, Empty, GetPairingDataRequest,
+        GetPairingDataResponse,
+    },
+    tls::TlsConfig,
+};
+use bytes::BufMut as _;
+use picky::{pem::Pem, x509::Csr};
+use picky_asn1_x509::{PublicKey, SubjectPublicKeyInfo};
+use ring::{
+    rand,
+    signature::{self, EcdsaKeyPair, KeyPair},
+};
+
+use rustls_pemfile as pemfile;
+use tonic::transport::Channel;
+
+type Result<T, E = super::Error> = core::result::Result<T, E>;
+
+pub struct Connected(PairingClient<Channel>);
+pub struct Unconnected();
+
+pub struct Client<T> {
+    inner: T,
+    tls: TlsConfig,
+    uri: String,
+    rune: String,
+    key: Vec<u8>,
+}
+
+impl Client<Unconnected> {
+    pub fn new<C>(creds: C) -> Result<Client<Unconnected>>
+    where
+        C: TlsConfigProvider + RuneProvider,
+    {
+        let tls = creds.tls_config();
+        let rune = creds.rune();
+        let key = tls
+            .clone()
+            .private_key
+            .ok_or(Error::BuildClientError("empty tls private key".to_string()))?;
+        Ok(Self {
+            inner: Unconnected(),
+            tls,
+            uri: crate::utils::scheduler_uri(),
+            rune,
+            key,
+        })
+    }
+
+    pub fn with_uri(mut self, uri: String) -> Client<Unconnected> {
+        self.uri = uri;
+        self
+    }
+
+    pub async fn connect(self) -> Result<Client<Connected>> {
+        let channel = tonic::transport::Endpoint::from_shared(self.uri.clone())?
+            .tls_config(self.tls.inner.clone())?
+            .tcp_keepalive(Some(crate::TCP_KEEPALIVE))
+            .http2_keep_alive_interval(crate::TCP_KEEPALIVE)
+            .keep_alive_timeout(crate::TCP_KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(true)
+            .connect_lazy();
+
+        let inner = PairingClient::new(channel);
+
+        Ok(Client {
+            inner: Connected(inner),
+            tls: self.tls,
+            uri: self.uri,
+            rune: self.rune,
+            key: self.key,
+        })
+    }
+}
+
+impl Client<Connected> {
+    pub async fn get_pairing_data(&self, session_id: &str) -> Result<GetPairingDataResponse> {
+        Ok(self
+            .inner
+            .0
+            .clone()
+            .get_pairing_data(GetPairingDataRequest {
+                session_id: session_id.to_string(),
+            })
+            .await?
+            .into_inner())
+    }
+
+    pub async fn approve_pairing(
+        &self,
+        session_id: &str,
+        node_id: &[u8],
+        device_name: &str,
+        restrs: &str,
```

Comment:
Can the restrictions argument here be dropped?

If we allow the attestation device to put whatever restrictions they want, the application requesting the pairing may too little access to fulfill their use-case. Can we make it like android permissions where the base permissions are all or nothing? A gl-client consumer can show the user the restrictions put on the app with the pairing request when they receive it in `get_pairing_data`. Then the pairing can be completed from the restrictions from the original pairing request on the server-side. 

Also, what do you think about letting the attestation device reject the pairing request? I'm thinking it might be helpful for the requesting application on the UX side if the user doesn't like the proposed restrictions

Reply:
The attestation side of the pairing process is stateless at the moment as in: We don't carry any application data form step to step but If it turns out that this would be easier for developers once people started to use it, I am absolutely happy to change this.